Partial fix of weapon special marked as inactive

DisherProject opened the pull request:
Related to #3178 

This should fix weapon specials with "active_on" and "[filter_self] formula=" marked as inactive in the tooltips. 

I couldn't fix backstab because i didn't know what to do :p let me explain why in this example:

![capture](https://user-images.githubusercontent.com/29802612/41206584-6b5bc7c8-6d06-11e8-942e-96e3c912a8de.PNG)

The Thief will have backstab enabled vs the Sergeant, but not vs the Troll Whelp. So what should the tooltip say in this case? That backstab is active because there is _at least_ one enemy who satisfies the skill's requirements? Or should it say something else?
Afaik such issue may occur with any weapon special having "[filter_opponent]" or "[filter_defender]" as filters.

Vultraz commented:
In that case the tooltip should say it is active since there is at least one scenario in which that is so.

CelticMinstrel commented:
Isn't there already logic in the engine to decide whether a special is active for the purpose of display? Something with `set_specials_context_for_listing()` or something?

DisherProject commented:
Ye actually `attack_type::special_context` is meant to determine whether to show the tooltip or not. Afaik it is right to call it there, but putting the call inside the scope produces an unexpected behaviour: that is, the unit is never marked as "attacking", because of the destructor of `special_context`.

DisherProject commented:
I've updated the commits: now the fix includes specials like backstab too. Please notice that backstab will now be evaluated as active even though there are no adjacent enemies.

CelticMinstrel commented:
I think I need to view the whitespace-ignoring diff to see what you've done here…

DisherProject commented:
In the first commit i removed the scopes around the specials contexts. 
As you can see in the context [destructor](https://github.com/wesnoth/wesnoth/blob/master/src/units/attack_type.cpp#L580), once the context [leaves](https://github.com/wesnoth/wesnoth/blob/master/src/reports.cpp#L813) the scope, the `attack_type` forgets everything about its relative unit etc... in particular it doesn't know anymore whether the unit is attacking or not.
As a consequence, when the `special_context_for_listing` is initialized you will only know that `is_for_listing_` is set to `true`.
Anytime the weapon special has the `active_on` key, the [code](https://github.com/wesnoth/wesnoth/blob/master/src/units/abilities.cpp#L959-L968) will try to see if the unit is actually attacking. 
(Probably you know better than me how this piece of code works, but anyway) given these preconditions, this last check will always fail.
So, what i did in the first commit was letting the context know if the unit was attacking.

In the second commit i have added the [new method](https://github.com/wesnoth/wesnoth/pull/3232/commits/c9ef0f649d4beec424c78cd0903ec0937ae53d83#diff-2d3ba6a50fd8a9952d1b9839cd0c4767L814) `special_context_t::set_specials_context_for_listing()` which simply allows you to set `is_for_listing_`. Then i used this new setter in the line i have linked.

To be honest, one could also initialize a new context which shared the `attack_type` with `ctx`, instead of creating a new method. Anyway, imo removing the scopes is necessary.

(I hope this is clear, as English is not my strong point)

CelticMinstrel commented:
Hmm, so couldn't this be done while retaining the scopes if you added an `attacking` parameter to `specials_context_for_listing()`?

DisherProject commented:
Yes of course, this can be done as you say.
If i have to add an `attacking` parameter while maintaining the scopes, i wouldn't even need `set_specials_context_for_listing` because i can implement it inside the `specials_context_for_listing` constructor.

Anyway, i didn't do this because i generally don't enjoy repeating the same code in different lines; also, imho adding more parameters to such a simple function could make it less readable. 
Being the `attack_type` a `const` i thought i could remove scopes, because it is read-only.

I'll do as you say anyway xD

DisherProject commented:
Here's the new commit. This is for master branch, so i can make another pull request for 1.14 if you want

CelticMinstrel commented:
I'll cherry-pick it to 1.14 soon.
